### PR TITLE
add information on the system type string

### PR DESCRIPTION
**Pull request by fricklerhandwerk**
# Motivation

This one in a series of small PRs to slightly restructure documentation on built-in functions, with the ultimate goal of providing a better overview on `builtins.derivation` and an overall more consistent presentation.

This work is sponsored by [Antithesis](https://antithesis.com/) ✨

# Context

Depends on https://github.com/NixOS/nix/pull/8314

This work is sponsored by [Antithesis](https://antithesis.com/) ✨
# Checklist for maintainers

Maintainers: tick if completed or explain if not relevant

 - [x] agreed on idea
 - [x] agreed on implementation strategy
 - [ ] tests, as appropriate
   - functional tests - `tests/**.sh`
   - unit tests - `src/*/tests`
   - integration tests - `tests/nixos/*`
 - [x] documentation in the manual
 - [x] documentation in the internal API docs
 - [ ] code and comments are self-explanatory
 - [ ] commit message explains why the change was made
 - [ ] new feature or incompatible change: updated release notes

# Priorities

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

### CODE REVIEW

**Comment by fricklerhandwerk** [comment]
Triaged in Nix team meeting 2023-05-19:

- @edolstra: this is probably the worst place to document it
    - also the system string is an implementation detail
    - it may be worthwhile knowing these things when porting Nix
    - probably having an allowlist would be better
        - @ericson2314: no, you want canonical hostnames to do cross compilation
            - also if you pass `--host`, `--build`, `--target` to `./configure` it will be [`config.sub`](https://git.savannah.gnu.org/cgit/config.git/tree/config.sub) that produces the string
    - there is also the [`system`](https://nixos.org/manual/nix/unstable/command-ref/conf-file.html#conf-system) setting
        - `builtins.currentSystem` should point to `system` as that is what it returns
        - should list the supported system types (systems Nix is ported to):
            - `x86_64-linux`
            - `x86_64-darwin`
            - `aarch64-linux`
            - `aarch64-darwin`
            - `i686-linux`
    - @ericson2314: really the place for this informatin is the `system` field in store derivations
- @fricklerhandwerk will move it to the `system` setting docs for now

**Comment by fricklerhandwerk** [comment]
@roberth please check this one, I implemented all the changes since last review.

File: doc/manual/src/contributing/hacking.md [hunk 1]
```diff
@@ -110,41 +110,72 @@ You can also build Nix for one of the [supported platforms](#platforms).
...
+```
+
+When Nix is built such that `./configure` is passed any of the `--host`, `--build`, `--target` options, the value is based on the output of [`config.sub`](https://github.com/nixos/nix/blob/master/config/config.sub) ([upstream](https://git.savannah.gnu.org/cgit/config.git/tree/config.sub)):
+
+```
+<cpu>-<vendor>[-<kernel>]-<os>
+```
 
-You can use any of the other supported platforms in place of `aarch64-linux`.
+For historic reasons and backward-compatibility, some CPU and OS identifiers are transformed as follows in [`configure.ac`](https://github.com/nixos/nix/blob/master/config/config.sub):
```
Comment (roberth) [inline doc/manual/src/contributing/hacking.md:169]: This sentence appears to assume that a standard naming exists, but this is relative to the GNU naming convention, as we use autotools.

```suggestion
For historic reasons and backward-compatibility, some CPU and OS identifiers are translated from the GNU naming convention as follows in [`configure.ac`](https://github.com/nixos/nix/blob/master/config/config.sub):
```

File: doc/manual/src/language/builtin-constants.md [hunk 2]
```diff
@@ -14,9 +14,8 @@ These constants are built into the Nix language evaluator:
 
 - [`builtins.currentSystem`]{#builtins-currentSystem} (string)
 
-  The built-in value `currentSystem` evaluates to the Nix platform
-  identifier for the Nix installation on which the expression is being
-  evaluated, such as `"i686-linux"` or `"x86_64-darwin"`.
+  Return the system type of the Nix installation on which the expression is being evaluated.
+  This value is determined by the [`system` configuration setting](@docroot@/command-ref/conf-file.md#conf-system).
```
Comment (roberth) [inline doc/manual/src/language/builtin-constants.md:18]: Problems:
 - `system` doesn't _have_ to be the system of the Nix installation.
 - values don't return anything.

```suggestion
  The system type of the Nix installation on which the expression is being evaluated, or the [`system` configuration setting](@docroot@/command-ref/conf-file.md#conf-system) if set to a different value.
```

Did I keep this in the order of relevance or did I make it contrived?
At least it's more truthful.

Comment (fricklerhandwerk) [reply doc/manual/src/language/builtin-constants.md:18]:
```suggestion
  The value of the [`system` configuration setting](@docroot@/command-ref/conf-file.md#conf-system).
```

File: doc/manual/src/contributing/hacking.md [hunk 3]
```diff
@@ -110,41 +110,72 @@ You can also build Nix for one of the [supported platforms](#platforms).
...
+on, you need a way for your current Nix installation to build code for that
+platform. Common solutions include [remote builders] and [binary format emulation]
 (only supported on NixOS).
 
 [remote builders]: ../advanced-topics/distributed-builds.md
 [binfmt emulation]: https://nixos.org/manual/nixos/stable/options.html#opt-boot.binfmt.emulatedSystems
 
-These solutions let Nix perform builds as if you're on the native platform, so
-executing the build is as simple as
+Given such a setup, executing the build only requires selecting the respective output attribute.
```
Comment (roberth) [inline doc/manual/src/contributing/hacking.md:133]:
```suggestion
Given such a setup, executing the build only requires selecting the respective attribute.
```

Not related to outputs. (Maybe flake outputs, unfortunately, but no flakes here technically..)

File: src/libstore/globals.hh [hunk 4]
```diff
@@ -193,18 +193,24 @@ public:
...
-          value from its default, since you can use it to ‘lie’ about the
-          platform you are building on (e.g., perform a Mac OS build on a
-          Linux machine; the result would obviously be wrong). It only makes
-          sense if the Nix binaries can run on multiple platforms, e.g.,
-          ‘universal binaries’ that run on `x86_64-linux` and `i686-linux`.
-
-          It defaults to the canonical Nix system name detected by `configure`
-          at build time.
+          The system type of the current Nix installation.
+          Nix will only build [derivations](@docroot@/language/derivations.md) whose `system` attribute equals the value specified here.
```
Comment (roberth) [inline src/libstore/globals.hh:197]:
```suggestion
          Nix will only build [derivations](@docroot@/language/derivations.md) locally when the `system` attribute equals the value specified here, or one of the [extra-platforms](@docroot@/command-ref/conf-file.html#conf-extra-platforms).
```

File: src/libstore/globals.hh [hunk 5]
```diff
@@ -193,18 +193,24 @@ public:
...
+
+          - `x86_64-linux`
+          - `x86_64-darwin`
+          - `i686-linux`
+          - `aarch64-linux`
+          - `aarch64-darwin`
+          - `armv6l-linux`
+          - `armv7l-linux`
+
+          In general, it never makes sense to modify this value when building derivations.
```
Comment (roberth) [inline src/libstore/globals.hh:210]:
```suggestion
          In general, you do not have to modify this setting.
```

File: src/libstore/globals.hh [hunk 6]
```diff
@@ -193,18 +193,24 @@ public:
...
+          While you can force Nix to run a Darwin-specific `builder` executable on a Linux machine, the result would obviously be wrong.
+
+          This value is available in the Nix language as [`builtins.currentSystem`](@docroot@/language/builtin-constants.md#builtins-currentSystem).
```
Comment (roberth) [inline src/libstore/globals.hh:213]: To be changed by
 - #4093
 
(for reference; no action needed now)

**Review by roberth — COMMENTED** [review]
Almost there

File: doc/manual/src/language/constructs.md [hunk 7]
```diff
@@ -217,18 +217,16 @@ three kinds of patterns:
     
     > **Warning**
     > 
-    > The `args@` expression is bound to the argument passed to the
-    > function which means that attributes with defaults that aren't
-    > explicitly specified in the function call won't cause an
-    > evaluation error, but won't exist in `args`.
-    > 
+    > `args@` binds the name `args` to the attribute set that is passed to the function.
```
Comment (fricklerhandwerk) [inline doc/manual/src/language/constructs.md:220]: Not sure how this got here. Will clean that up.

File: src/libstore/globals.hh (hunk 4 above)
Comment (fricklerhandwerk) [reply src/libstore/globals.hh:197]:
```suggestion
          Nix will only build a given [derivation](@docroot@/language/derivations.md) locally when its `system` attribute equals the value specified here or one of the [`extra-platforms`](@docroot@/command-ref/conf-file.html#conf-extra-platforms).
```

**Review by roberth — APPROVED** [review]

**Pull request merged**